include/sensors/core/fdc1004.hpp

```diff
@@ -27,39 +27,448 @@
+    // b10 200S/s
+    // b11 400S/s
+    RESERVED = 0x0,
+    ONE_HUNDRED_SAMPLES_PER_SECOND = 0x1,
+    TWO_HUNDRED_SAMPLES_PER_SECOND = 0x2,
+    FOUR_HUNDRED_SAMPLES_PER_SECOND = 0x3,
+};
+
+// This is used to determine which configuration register
+// you're currently using and thus what measurement mode
+// you're in.
+enum class MeasureConfigMode : uint8_t { ONE, TWO, THREE, FOUR };
+
+// Capacitance Sensor Address and Registers
+enum class Registers : uint8_t {
+    MEAS1_MSB = 0x0,
+    MEAS1_LSB = 0x01,
+    MEAS2_MSB = 0x02,
+    MEAS2_LSB = 0x03,
+    MEAS3_MSB = 0x04,
+    MEAS3_LSB = 0x05,
+    MEAS4_MSB = 0x06,
+    MEAS4_LSB = 0x07,
+    CONF_MEAS1 = 0x08,
+    CONF_MEAS2 = 0x09,
+    CONF_MEAS3 = 0x0A,
+    CONF_MEAS4 = 0x0B,
+    FDC_CONF = 0x0C,
+    OFFSET_CAL_CIN1 = 0x0D,
+    OFFSET_CAL_CIN2 = 0x0E,
+    OFFSET_CAL_CIN3 = 0x0F,
+    OFFSET_CAL_CIN4 = 0x10,
+    GAIN_CAL_CIN1 = 0x11,
+    GAIN_CAL_CIN2 = 0x12,
+    GAIN_CAL_CIN3 = 0x13,
+    GAIN_CAL_CIN4 = 0x14,
+    MANUFACTURER_ID = 0xFE,
+    DEVICE_ID = 0xFF
+};
+
+static inline auto is_valid_address(const uint8_t add) -> bool {  // NOLINT
+    switch (static_cast<Registers>(add)) {
+        case Registers::MANUFACTURER_ID:
+        case Registers::DEVICE_ID:
+        case Registers::MEAS1_MSB:
+        case Registers::MEAS1_LSB:
+        case Registers::MEAS2_MSB:
+        case Registers::MEAS2_LSB:
+        case Registers::MEAS3_MSB:
+        case Registers::MEAS3_LSB:
+        case Registers::MEAS4_MSB:
+        case Registers::MEAS4_LSB:
+        case Registers::CONF_MEAS1:
+        case Registers::CONF_MEAS2:
+        case Registers::CONF_MEAS3:
+        case Registers::CONF_MEAS4:
+        case Registers::FDC_CONF:
+        case Registers::OFFSET_CAL_CIN1:
+        case Registers::OFFSET_CAL_CIN2:
+        case Registers::OFFSET_CAL_CIN3:
+        case Registers::OFFSET_CAL_CIN4:
+        case Registers::GAIN_CAL_CIN1:
+        case Registers::GAIN_CAL_CIN2:
+        case Registers::GAIN_CAL_CIN3:
+        case Registers::GAIN_CAL_CIN4:
+            return true;
+    }
+    return false;
+}
+
+/** Template concept to constrain what structures encapsulate registers.*/
+template <typename Reg>
+// Struct has a valid register address
+// Struct has an integer with the total number of bits in a register.
+// This is used to mask the value before writing it to the sensor.
+concept FDC1004Register =
+    std::same_as<std::remove_cvref_t<decltype(Reg::address)>,
+                 std::remove_cvref_t<Registers&>> &&
+    std::integral<decltype(Reg::value_mask)>;
+
+template <typename Reg>
+concept WritableRegister = requires() {
+    {Reg::writable};
+};
+
+template <typename Reg>
+concept ReadableRegister = requires() {
+    {Reg::readable};
+};
+
+struct __attribute__((packed, __may_alias__)) ManufactureId {
+    static constexpr Registers address = Registers::MANUFACTURER_ID;
+    static constexpr bool readable = true;
+    static constexpr bool writable = false;
```

Comment:
The way that the `ReadableRegister` and `WritableRegister` concepts work is extremely unintuitive, but it actually evaluates whether this variable _exists_ rather than looking at the value of the constexpr (which isn't accessible in concept evaluation). So any of these registers with `readable = false` or `writable = false` should just have that respective struct member removed. Sorry!